src/Compilers/CSharp/Portable/BoundTree/BoundCollectionExpression.cs

```diff
@@ -3,19 +3,45 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
     internal partial class BoundCollectionExpression
     {
         private partial void Validate()
         {
+#if DEBUG
+            var collectionCreation = this.CollectionCreation;
+            while (collectionCreation is BoundConversion conversion)
+                collectionCreation = conversion.Operand;
+
+            Debug.Assert(collectionCreation
+                            is null
+                            or BoundObjectCreationExpression
+                            or BoundCall
+                            or BoundNewT
+                            or BoundBadExpression);
+
+            if (collectionCreation != null &&
+                this.HasWithElement)
+            {
+                Debug.Assert(collectionCreation.Syntax is WithElementSyntax);
```

Comment:
>collectionCreation.Syntax is WithElementSyntax [](http://example.com/codeflow?start=29&length=46)

We generally do not restrict syntax for bound nodes. Its value is not supposed to have any semantical meaning, i.e. shouldn't affect how the node is interpreted. For example, Compiler generated nodes should be able to use any value.